### PR TITLE
[V0][Sampler] Use raw logits for greedy argmax

### PR DESCRIPTION
**Pull request by njhill**
To hopefully avoid some of the reported precision-related nondeterminism.

Also delete a vestigial intermediate method.

### CODE REVIEW

**Comment by patrickvonplaten** [comment]
Is this fix not necessary for the v1 sampler?

**Comment by njhill** [comment]
> Is this fix not necessary for the v1 sampler?

@patrickvonplaten I don't think so, we were already using logits rather than logprobs for this in V1: https://github.com/vllm-project/vllm/blob/9f37422779a17d8b2ebde300808166068f4ad4cc/vllm/v1/sample/sampler.py#L77-L78

**Comment by tonyaw** [comment]
@njhill, may I ask when this PR will be merged? Any target release? Thanks in advance!

**Comment by njhill** [comment]
@tonyaw there are test failures that need investigating: https://buildkite.com/vllm/ci/builds/15883#0195b060-d27c-4bd5-b435-c495d3709d24, any help would be welcome!

**Comment by gx16377** [comment]
@njhill Hi, may I ask why argmax(logits) is more stable than argmax(logprobs)?

**Comment by tonyaw** [comment]
> @njhill Hi, may I ask why argmax(logits) is more stable than argmax(logprobs)?

@njhill, I also want to know the reason, could you please help to explain? :-)

**Comment by njhill** [comment]
@gx16377 @tonyaw logprob is softmax of the logits which is a nonlinear projection of the entire floating point range into the range [0,1]. So it essentially reduces precision and many token values will end up tied that weren't beforehand, including values tied for first place, and argmax may select from these arbitrarily (e.g. could vary by batch size).

**Comment by gx16377** [comment]
> @gx16377 @tonyaw logprob is softmax of the logits which is a nonlinear projection of the entire floating point range into the range [0,1]. So it essentially reduces precision and many token values will end up tied that weren't beforehand, including values tied for first place, and argmax may select from these arbitrarily (e.g. could vary by batch size).

Thank you

**Comment by tonyaw** [comment]
> > @gx16377 @tonyaw logprob is softmax of the logits which is a nonlinear projection of the entire floating point range into the range [0,1]. So it essentially reduces precision and many token values will end up tied that weren't beforehand, including values tied for first place, and argmax may select from these arbitrarily (e.g. could vary by batch size).
> 
> Thank you
Thanks!
Could you please explain more about "could vary by batch size"?
Does "batch size" here mean concurrent request count or sth else? 
If it is "concurrent request count", could you please explain more why it is related to softmax calculation?

**Comment by tonyaw** [comment]
@njhill Thanks!
In vllm, all the calculation is done by matrix.
When multiple requests are processed together, logits of different requests are put into one matrix, so it is possible that calculation of logits of different requests may impact each other.
Is my understanding right? :-)

**Comment by saikrishb** [comment]
> I mean argmax. May not be deterministic when there are tied max values.

@njhill [torch documentation](https://pytorch.org/docs/stable/generated/torch.argmax.html) says argmax will pick the lowest index. At least for the case of multiple max logits, this commit will make sampling deterministic instead of picking one of the max values as mentioned [here](https://github.com/vllm-project/vllm/issues/5404#issuecomment-2660528911) correct?

**Review by aarnphm — APPROVED** [review]
